src/journal/src/Journal/Internal/ByteBuffer.hs

```diff
@@ -0,0 +1,306 @@
+{-# LANGUAGE UnboxedTuples #-}
+
+module Journal.Internal.ByteBuffer where
+
+import Control.Exception
+import Data.IORef
+import Foreign
+import GHC.ForeignPtr
+import GHC.Exts
+import GHC.Types
+import GHC.IO
+import System.Posix.IO (openFd, defaultFileFlags, OpenMode(ReadWrite))
+
+import Journal.Internal.Mmap
+
+------------------------------------------------------------------------
+-- * Types
+
+type MBArray = MutableByteArray# RealWorld
+
+data ByteBuffer = ByteBuffer
+  { bbData     :: {-# UNPACK #-} !MBArray
+  , bbCapacity :: {-# UNPACK #-} !Capacity
+  , bbLimit    :: {-# UNPACK #-} !(IORef Limit)
+  , bbPosition :: {-# UNPACK #-} !(IORef Position)
+  , bbMark     :: {-# UNPACK #-} !(IORef Position)
+  , bbSlice    :: {-# UNPACK #-} !(IORef Slice)
+  }
+
+newtype Capacity = Capacity Int
+  deriving (Num, Integral, Real, Ord, Eq, Enum)
+
+newtype Limit = Limit Int
+  deriving (Num, Integral, Real, Ord, Eq, Enum)
+
+newtype Position = Position Int
+  deriving (Num, Integral, Real, Ord, Eq, Enum)
+
+newtype Slice = Slice Int
+  deriving (Num, Integral, Real, Ord, Eq, Enum)
+
+------------------------------------------------------------------------
+
+newByteBuffer :: MBArray -> Capacity -> Limit -> Position -> Maybe (IORef Slice)
+              -> IO ByteBuffer
+newByteBuffer mba# capa lim pos mSli
+  = ByteBuffer mba# capa
+  <$> newIORef lim
+  <*> newIORef pos
+  <*> newIORef (-1)
+  <*> maybe (newIORef 0) return mSli
+
+bbPtr :: ByteBuffer -> Ptr a
+bbPtr (ByteBuffer mba# _ _ _ _ _) = Ptr (byteArrayContents# (unsafeCoerce# mba#))
+{-# INLINE bbPtr #-}
+
+getCapacity :: ByteBuffer -> Capacity
+getCapacity = bbCapacity
+{-# INLINE getCapacity #-}
+
+readLimit :: ByteBuffer -> IO Limit
+readLimit = readIORef . bbLimit
+{-# INLINE readLimit #-}
+
+writeLimit :: ByteBuffer -> Limit -> IO ()
+writeLimit bb = writeIORef (bbLimit bb)
+{-# INLINE writeLimit #-}
+
+readPosition :: ByteBuffer -> IO Position
+readPosition = readIORef . bbPosition
+{-# INLINE readPosition #-}
+
+writePosition :: ByteBuffer -> Position -> IO ()
+writePosition bb = writeIORef (bbPosition bb)
+{-# INLINE writePosition #-}
+
+incrPosition :: ByteBuffer -> Int -> IO ()
+incrPosition bb i = modifyIORef (bbPosition bb) (+ fromIntegral i)
+{-# INLINE incrPosition #-}
+
+readMark :: ByteBuffer -> IO Position
+readMark = readIORef . bbMark
+{-# INLINE readMark #-}
+
+writeMark :: ByteBuffer -> Position -> IO ()
+writeMark bb = writeIORef (bbMark bb)
+{-# INLINE writeMark #-}
+
+readSlice :: ByteBuffer -> IO Slice
+readSlice = readIORef . bbSlice
+
+writeSlice :: ByteBuffer -> Slice -> IO ()
+writeSlice bb = writeIORef (bbSlice bb)
+
+------------------------------------------------------------------------
+
+remaining :: ByteBuffer -> IO Int
+remaining bb = do
+  lim <- readLimit bb
+  pos <- readPosition bb
+  return (fromIntegral (lim - fromIntegral pos))
+
+------------------------------------------------------------------------
+-- * Checks
```

Comment:
Hmm don't have any particular suggestion here, but I wonder if we can get something closer to "parse don't validate". https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

Reply:
We have a parse function in `Journal.Internal.Parse`, but that will build an AST and therefore do copying (or at least allocations), which is fine for static analysis to ensure that the log is consistent, e.g. at start up. But I think bound checks will need to be done at run-time (unless we introduce some fancy types, which I doubt we want)? We should probably add unsafe versions of the functions using bound checks which skips the checks, but then segfaults can happen... As for the invariant, it could be useful for property testing (and maybe moved to that module once it exists).